File: Front/ios/Petmily/KakaoMapViewManager.m
#import <React/RCTViewManager.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTUIManager.h>

// ViewManager 등록
// Swift 클래스는 @objc(KakaoMapViewManager)로 선언되어 있으므로
// Objective-C에서 접근 가능합니다
@interface RCT_EXTERN_MODULE(KakaoMapViewManager, RCTViewManager)

// React Native 속성들
RCT_EXPORT_VIEW_PROPERTY(apiKey, NSString)
RCT_EXPORT_VIEW_PROPERTY(latitude, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(longitude, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(zoomLevel, NSNumber)

// 명령 메서드
RCT_EXTERN_METHOD(addMarker:(nonnull NSNumber *)node
                  latitude:(nonnull NSNumber *)latitude
                  longitude:(nonnull NSNumber *)longitude
                  title:(nonnull NSString *)title)

RCT_EXTERN_METHOD(updateRoute:(nonnull NSNumber *)node
                  routeID:(nonnull NSString *)routeID
                  coordinates:(nonnull NSArray *)coordinates
                  colorHex:(nonnull NSNumber *)colorHex
                  strokeColorHex:(nonnull NSNumber *)strokeColorHex
                  lineWidth:(nonnull NSNumber *)lineWidth)

RCT_EXTERN_METHOD(clearRoute:(nonnull NSNumber *)node
                  routeID:(nonnull NSString *)routeID)

RCT_EXTERN_METHOD(startCameraTracking:(nonnull NSNumber *)node)

RCT_EXTERN_METHOD(stopCameraTracking:(nonnull NSNumber *)node)

RCT_EXTERN_METHOD(updateTrackingLocation:(nonnull NSNumber *)node
                  latitude:(nonnull NSNumber *)latitude
                  longitude:(nonnull NSNumber *)longitude)

@end

File: Front/ios/Petmily/KakaoMapViewManager.swift
import Foundation
import React // RCTBridge, RCTUIManager, RCTViewManager

@objc(KakaoMapViewManager)
class KakaoMapViewManager: RCTViewManager {

  override static func requiresMainQueueSetup() -> Bool { 
    return true 
  }

  // KakaoMapView를 생성해 RN에 제공
  override func view() -> UIView! {
    return KakaoMapView()
  }
  
  /// JS에서 보내는 명령: addMarker
  /// - Parameters:
  ///   - node: ReactTag (UIView의 태그)
  ///   - latitude/longitude: 좌표
  ///   - title: 마커 제목
  @objc
  func addMarker(_ node: NSNumber,
                 latitude: NSNumber,
                 longitude: NSNumber,
                 title: NSString) {
    DispatchQueue.main.async { [weak self] in
      guard
        let self = self,
        let uiManager = self.bridge?.uiManager,
        let component = uiManager.view(forReactTag: node) as? KakaoMapView
      else {
        return
      }

      component.addMarker(latitude, longitude: longitude, title: title)
    }
  }
  
  @objc
  func updateRoute(_ node: NSNumber,
                   routeID: NSString,
                   coordinates: NSArray,
                   colorHex: NSNumber,
                   strokeColorHex: NSNumber,
                   lineWidth: NSNumber) {
    DispatchQueue.main.async { [weak self] in
      guard
        let self = self,
        let uiManager = self.bridge?.uiManager,
        let component = uiManager.view(forReactTag: node) as? KakaoMapView
      else {
        return
      }
      
      component.updateRoute(routeID, coordinates: coordinates, colorHex: colorHex, strokeColorHex: strokeColorHex, lineWidth: lineWidth)
    }
  }
  
  @objc
  func clearRoute(_ node: NSNumber,
                  routeID: NSString) {
    DispatchQueue.main.async { [weak self] in
      guard
        let self = self,
        let uiManager = self.bridge?.uiManager,
        let component = uiManager.view(forReactTag: node) as? KakaoMapView
      else {
        return
      }
      
      component.clearRoute(routeID)
    }
  }
  
  @objc
  func startCameraTracking(_ node: NSNumber) {
    DispatchQueue.main.async { [weak self] in
      guard
        let self = self,
        let uiManager = self.bridge?.uiManager,
        let component = uiManager.view(forReactTag: node) as? KakaoMapView
      else {
        return
      }
      
      component.startCameraTracking()
    }
  }
  
  @objc
  func stopCameraTracking(_ node: NSNumber) {
    DispatchQueue.main.async { [weak self] in
      guard
        let self = self,
        let uiManager = self.bridge?.uiManager,
        let component = uiManager.view(forReactTag: node) as? KakaoMapView
      else {
        return
      }
      
      component.stopCameraTracking()
    }
  }
  
  @objc
  func updateTrackingLocation(_ node: NSNumber,
                              latitude: NSNumber,
                              longitude: NSNumber) {
    DispatchQueue.main.async { [weak self] in
      guard
        let self = self,
        let uiManager = self.bridge?.uiManager,
        let component = uiManager.view(forReactTag: node) as? KakaoMapView
      else {
        return
      }
      
      component.updateTrackingLocation(latitude, longitude: longitude)
    }
  }
}
